### ISSUE TITLE
Introducing NewPlayer

### ISSUE THREAD

**Issue by theScrabi** [TeamNewPipe/NewPipe#11603]
Dear NewPipers,

I give to you the fruit of my labor for the last three months that I worked for the [NewPipe Association](https://newpipe-ev.de): NewPipe's (hopefully) next player: NewPlayer.

But let's start from the beginning. NewPipe's current player implementation isn't very good. Its code is ugly, and it is based on the outdated ExoPlayer 2. So because of the ongoing refactoring process, I decided to rewrite the player and tackle the pain points we currently have with it.

<p align="center">
<img src="https://github.com/user-attachments/assets/d631a306-7d0a-49c8-839f-d2d4551e409f">
</p>

<h2 align="center"><b>NewPlayer</b></h3>

<h3 align="center">NewPipe's next media player framework</h3>

<br>
<br>
<i>Icon by <a href="https://jaim3.com">Jaime López</a></i>
<br>

### So what is NewPlayer then?
NewPlayer is a media framework independent of NewPipe itself. I decided to make it independent, because one of the big issues we have with the current player is that it is deeply integrated into NewPipe. Therefore, I wanted to make NewPlayer a separate module in order to enforce that the interface between NewPipe and the player is only as big as necessary. This also has the advantage that NewPlayer can be used independently of NewPipe itself, which means it can be used in other apps too.

### Let me give you a Profile about NewPlayer:
- It is a module, separate from NewPipe and can be used as an independent player framework
- It is based on the [Media3](https://developer.android.com/media/media3) library
- Its UI is created with [Jetpack Compose](https://developer.android.com/compose)
- It's fully written in Kotlin
- Its UI resembles the UI of the current NewPipe player, but improves it
- The UI uses [Material You](https://m3.material.io/blog/announcing-material-you) theming
- It follows an [MVVM](https://www.geeksforgeeks.org/mvvm-model-view-viewmodel-architecture-pattern-in-android/) architecture
- It is GPLv3 licensed

### How does NewPlayer look like?
<table>
<tr>
<td>
<img width="200" src="https://github.com/TeamNewPipe/NewPlayer/raw/master/misc/screenshots/373686583-1164cf7c-66eb-48be-aeda-55e6e6294cf1.png">
<br>
Embedded Screen in test App
</td>
<td><img width="550" src="https://github.com/TeamNewPipe/NewPlayer/raw/master/misc/screenshots/373685724-42609e51-6bf7-4008-b084-a59ce111f3c1.png">
<br>
Fullscreen
</td>
</tr>
</table>
<table>
<tr>
<td>
<img width="350" src="https://github.com/TeamNewPipe/NewPlayer/raw/master/misc/screenshots/373688583-9011749c-3aec-4bf7-a368-40000c84f8e3.png">
<br>
Audio frontend
</td>
<td>
<img width="400" src="https://github.com/TeamNewPipe/NewPlayer/raw/master/misc/screenshots/373689058-9fc27dfd-7f89-48de-b0ff-cd6e9bd4fdbd.png">
<br>
Audio frontend landscape
</td>
</tr>
</table>

<table>
<tr>
<td>
<img width="500" src="https://github.com/TeamNewPipe/NewPlayer/raw/master/misc/screenshots/373690788-a7af6db0-eac2-4913-8f60-3d621c5afd9f.png">
<br>
Playlist screen
</td>
<td>
<img width="500" src="https://github.com/TeamNewPipe/NewPlayer/raw/master/misc/screenshots/373690876-9b96ae22-d537-4b49-ac1c-4549a94bebcb.png">
<br>
Chapter screen
</td>
<td>
<img width="500" src="https://github.com/TeamNewPipe/NewPlayer/raw/master/misc/screenshots/373691456-4aaff87d-dbf8-4877-866b-60e6fc05ea6a.png">
<br>
Picture in Picture
</td>
</tr>
</table>
<table>
<tr>
<td>
<img width="400" src="https://github.com/TeamNewPipe/NewPlayer/raw/master/misc/screenshots/373692488-5e861e22-a969-4eae-aa05-ecd9a339e80d.png">
<br>
Volume indicator
</td>
<td>
<img width = "400" src="https://github.com/TeamNewPipe/NewPlayer/raw/master/misc/screenshots/373695908-341112d4-dac0-488f-961c-9b389396d289.png">
<br>
Main menu
</td>
</tr>
</table>

### Here a video showing NewPlayer

https://github.com/user-attachments/assets/6e128b5b-184b-429c-baf6-886eab153854

## Try It your self
<h3><a href="https://cloud.newpipe-ev.de/s/lHLwwbqlmOvZhqh">Download</a> NewPlayer testapp from NewPipe e.V.</h3>

### [Here](https://github.com/TeamNewPipe/NewPlayer) you can find the code of NewPlayer

# Things that still need to be done
Well. NewPlayer is not yet finished. It still has many quirks and there are pieces that are straight up missing. However, I want to start to slowly hand over the development, as I will start a 9 to 5 job at the end of this month. So after this month I will only be able to give advice or feedback, but I will not really be able to actively develop anymore.

Here is a list of things that are still missing or are not correctly working yet. If you want to help, please note I have sorted the list by importance. I will work on the most important things by myself. So if you want to help, you might not want to pick the very first item, in order for us not collide with our development efforts.

### Things that still need to be done
- [ ] **Stream and Track selection**
- [ ] **Decide weather to call `NewPlayer` `NewPlayer` or `NewPlay`**
- [x] **Documentation**
  Documentation is critical. I will do most of it myself before I leave for my new job. This way you can better understand how to use NewPlayer or what my intentions were.
- [ ] **UI for Speed and Pitch selection**
- [x] **Content ratio not correct in embedded view**
  Sometimes the video is not correctly fitted within the embedded view. The video itself is stretched while black bars are visible on the top and the bottom. The black bars should not be there, and the video should not be stretched vertically.
- [x] **Fix Content fit mode**
  Changing the content fit mode is not working correctly, and the menu item does nothing atm.
- [ ] **Default brightness can not be restored when leaving fullscreen mode**
- [ ] **Original rotation is not always restored correctly when leaving fullscreen mode**
- [ ] **Playback mode changes made from outside NewPlayer can introduce faulty playback states**
- [ ] **System bar color can not be restored when leaving fullscreen mode**
- [ ] **Subtitle selection**
  Subtitle selection and subtitle settings is completely missing atm.
- [ ] **Proper layout**
  The layout of many UI elements in NewPlayer are only roughly placed. My GF is a multimedia designer. She didn't like the UI, but offered to help place the elements better.
- [ ] **Proper theming (required by Audio player UI)**
  The theme atm is only sort of a placeholder. It is not final and should therefore be adjusted to match the colors for NewPipe. Even more there should be a way to customize the theme without having to recompile the whole player. This could be achieved by making the theme be a parameter of the `NewPlayer` class constructor.
  Furthermore, NewPlayer currently has no Light theme. This was important for the audio UI.
- [ ] **Testing**
  There are no tests for NewPlayer. I simply did not have the time yet. However, it would be good to have some UI and unit tests. Additionally, there should be fuzzy tests, because simply tapping randomly on the UI still sometimes breaks the player.
- [x] **Auxiliary repositories**
  I imagined 3 auxiliary repository implementations that can be used in combination with an actual repository.
  These are:
    - `MultiRepo`: Helps to combine multiple Repositories into one repo. This way each actual repo can focus on serving one backend.
    - `CachingRepo`: A repository that helps as an in memory cache. This is necessary since NewPlayer does not provide a cache itself and basically requests the same Data over and over again if it needs it. (@TeamNewPipe/appdev: Don't use this! NewPipe has its own cache. It would be better if NewPipe used its already existing cache, so data can be shared between NewPipe and NewPlayer)
    - `PrefetchRepo`: A repo that requests all possible requests it can make once it sees a new `item`. This can be used in combination with `CachingRepo` to ensure that the caches are eagerly filled upon loading a new `item`.
- [x] **Animations are sometimes missing or are stuttery or clunky**
- [ ] **Integration with NewPipe**
- [ ] **DPad Support**
  The current player of NewPipe is not usable with only a DPad and an Enter key. This makes NewPipe unusable on a TV. Now that we rewrite the player we could try to put more attention at this. The issue now is that Jatpack Compose does not seem to be very good at supporting DPad input: see [here](https://github.com/thesauri/dpad-compose)
- [ ] **TV Mode**
  TV Mode is a special mode for NewPlayer, intended for Android TV devices. You see, by default Media3 devs suggest putting the ExoPlayer into a MediaSession Service. However, NewPlayer puts the ExoPlayer instance into its `NewPlayer` object, which again lives in the `Application` instance. Therefore, NewPlayer can be used without a MediaSession and its respective foreground Service. This would be beneficial for Android TV devices, as these often don't support Notifications. This then means, if I close NewPipe on, let's say, a FireTV Stick, it would continue to playback in background. However, you won't be able to control the playback unless you open the App again. The actual expected behavior should be that playback stops, which would work if the MediaSession was disabled. The TV Mode, and the DPad support would be necessary for NewPlayer to support Android TV.
- [ ] **Clipped text does not scroll**
  Some text, like titles or chapter descriptions are too large to fit in the screen. NewPipe currently scrolls through such text, so the user is able to read the entire text. NewPlayer ellipses the text atm. It was nice however if the text was scrolling again.

**Comment by acrodemocide** [comment]
@opusforlife2 Yes, I'm interested in contributing. I've mostly reviewed pull requests and have tested the builds on my Android device as part of the review to let report back whether it's working or not.

If I have the time, I'll look into picking up some issues to make code changes for, but in the last update, the request was mostly for support in code reviews and testing, and I'm happy to keep doing that in my free time. I love NewPipe and want to see it continue to be successful

**Comment by toliuweijing** [comment]
Yes. Should be able to pick up items after understanding the NewPlayer codebase. 

I'm new to open source contribution. Sometime I found myself need support / discussion to unblock myself. Is there a preferable way to discuss the refactoring work? e.g. a dedicated chat room for people active on the refactoring work should be perfect. cc @theScrabi @opusforlife2 WDYT

**Comment by nirodhvana** [comment]
Cant we just design the extra player options(subtitles,etc) like it is right now instead of a triple dot menu 

it is much easier to navigate them

**Comment by Stypox** [comment]
@toliuweijing yes, we have a chat room, see https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md#communication

**Comment by gigigigi53** [comment]
nice

**Comment by zeropi2021** [comment]
Sounds very interesting but doesn't work on Android 7. The app tries to start but crashes.

**Comment by theScrabi** [comment]
Hei @toliuweijing, I wrote a bugreport over at the NewPlayer repo: https://github.com/TeamNewPipe/NewPlayer/issues/3

Maybe you can already have a look at it. There are three levels you can go:
1. Try to reproduce the bug that I describe
2. Find the code that causes that issue
3. Fix the issue and send a PR

Would you want to try it out?

**Comment by jm355** [comment]
What do you think about being able to use newplayer to play local video files?

**Comment by theScrabi** [comment]
Sure, but someone else would have to make NewPlayer do that. 
NewPlayer is written like a framework. So anyone can feel free to write an app around NewPlayer to make it able to play back local files.

**Comment by jm355** [comment]
Ah I see, that makes sense. Hopefully the newpipe devs add that functionality then at some point! 🤞

**Comment by ccateni** [comment]
> Sure, but someone else would have to make NewPlayer do that. NewPlayer is written like a framework. So anyone can feel free to write an app around NewPlayer to make it able to play back local files.

Honestly, I would kill for an app that plays local files with the speed, pitch and skip silence controls of newpipe.

**Comment by theScrabi** [comment]
I put online parts of NewPlayer's documentation yesterday. Checkout: https://github.com/TeamNewPipe/NewPlayer
With this you should be able to experiment with NewPlayer for a bit :)

**Comment by theScrabi** [comment]
Oh, and by the way we now have seek preview:

https://github.com/user-attachments/assets/26527310-0647-43b6-9f73-3d3d2d97cf6d

**Comment by clemencyworld1** [comment]
Its a crashware for my phone. Nokia 2.3 running android 11
